[IR TX RX] Update the library and the Plugins

jimmys01 opened the pull request.

jimmys01 committed:
[IR TX RX] Update IR Library

jimmys01 committed:
[IR TX RX] Add the new protocols to the TX RX plugins

Also shorth alphabetically the protocols to match the library shorting

jimmys01 committed:
Small Fix

TD-er commented:
I've looked in the code of that plugin to see how `Plugin_035_irSender` is begin used.
For starters, it is never initialized to `nullptr` and it is also hardly anywhere being checked before it is de-referenced (calling the `->` on it)
Only in the write function is a single check on it and in the init function.
But if the pointer is never initialized to `nullptr`, there is no guarantee it will always work.

jimmys01 committed:
[IR] Latest Ir lib commit

Daikin AC related fix

TD-er commented on lib/IRremoteESP8266/src/ir_Daikin.h:
```diff
@@ -128,6 +128,7 @@ const uint64_t kDaikinFirstHeader64 =
     0b1101011100000000000000001100010100000000001001111101101000010001;
 
 // Another variant of the protocol for the Daikin ARC477A1 remote.
+const uint16_t kDaikin2Freq = 36700;  // Modulation Frequency in Hz.
```
Why are these made a global defined const variable, only to be used in a define?
This is taking more memory and it is confusing.

jimmys01 replied:
For this we should ask the library maintainer  
ref: markszabo/IRremoteESP8266#635

crankyoldgit replied:
AFAIUI, a `const` doesn't get allocated _space_ by the compiler unless it is needed/used. Similar to a `#define`, if it gets used somewhere, it then gets referenced. Hence it's similar to a `#define` (as in, a macro) but the compiler knows what type it is and _should_ use only one reference to the value if it is used multiple times. Also, `const`s containing equations are calculated at compile-time only as they naturally can only contain constants.
See also: https://google.github.io/styleguide/cppguide.html#Use_of_const

again, AFAIUI:
```
#define FOO 0x12345
do_something(FOO);
```
and
```
const uint16_t kFoo = 0x12345;
do_something(kFoo); 
```
and
```
do_something(0x12345); 
```
i.e. Somewhere the value of 0x12345 will ultimately be stored in the compiled binary.

Given a smart enough compiler (and Gnu's ones typically are) that should produce identical code and memory usage. The difference is, for the middle one, the compiler has clear type context and can produce better diagnostics.
e.g. Generate a warning when the value is larger than the type can hold.

If I'm wrong, please let me know.

P.S. If you want me to see a reference, please use a @crankyoldgit ;-) I only just noticed this when pushing out the v2.5.6 release of the library.

TD-er replied:
@crankyoldgit 
Well we may need to check this.
But as far as I know, these variables will be made available as an object on the system stack.
So they will use memory on the system stack (Arduino space is using 'cont' stack, the core libraries have the system stack)
It may very wel be that the compiler does see it is not referenced anywhere else, so it will then try to do some magic and not allocate it as an object on the stack, but to be sure, that has to be checked.
I don't think it will make much of a difference (if any) in the compiled binary size, but I'm more worried about the stack usage, since we've seen lots of strange errors in the last few months that appeared to be related to either one of both stacks.

The type checking is a good thing indeed.

crankyoldgit replied:
@TD-er I think the point is, you _shouldn't_ think of them as **variables** because they are _not_, They are **constants**, hence they are treated differently. Just like any other [literal](https://en.wikipedia.org/wiki/Literal_(computer_programming)) you'd use in the code. e,g `0x12345` in my example.

But yes. It needs to be confirmed. After all, it's the implementation that matters in the end.

TD-er replied:
A constant is not the same as a literal.

If you use a define, then the (pre)compiler will exchange it as if it is just code duplication in the source code.
If you create an object and reference it, then there will be an object which does use resources at runtime, like some allocation on the stack. (unless it is removed due to optimizations)
For a simple int, it will probably not matter much, but for more complex objects it does matter.

In the code, the define is referencing the actual variable, so at all places where that define is being used, the code has a reference to that object.

jimmys01 committed:
[IR TX RX] Add comments, Removed debug log prints, some refactoring to compact the code.

jimmys01 committed:
[IR RX] Make the log output more readable

TD-er commented on src/_P035_IRTX.ino:
```diff
@@ -353,12 +337,13 @@ boolean Plugin_035(byte function, struct EventStruct *event, String& string)
 
 boolean addErrorTrue() {
-    addLog(LOG_LEVEL_ERROR, F("RAW2: Invalid encoding!"));
+    addLog(LOG_LEVEL_ERROR, String(F("RAW2: Invalid encoding!")));
```
Why do you need this extra String wrapper?

jimmys01 replied:
in this cases like those, if you do not wrap it, it wont compile.
`addLog(LOG_LEVEL_ERROR, String(F("RAW2: Invalid encoding!")) + "ABCD");`
In the example you wrote about the `String `is not nessesary

jimmys01 commented:
Travis is not happy with P054
```
/home/travis/build/letscontrolit/ESPEasy/src/_P054_DMX512.ino: In function 'boolean Plugin_054(byte, EventStruct*, String&)':
/home/travis/build/letscontrolit/ESPEasy/src/_P054_DMX512.ino:250:26: error: call of overloaded 'write(int)' is ambiguous
Serial1.write(0);   //start byte
```

TD-er commented:
Ah, that's one of the more recent changes in the core 2.6.0 firmware, right?

jimmys01 commented:
> 
> 
> I've looked in the code of that plugin to see how `Plugin_035_irSender` is begin used.
> For starters, it is never initialized to `nullptr` and it is also hardly anywhere being checked before it is de-referenced (calling the `->` on it)
> Only in the write function is a single check on it and in the init function.
> But if the pointer is never initialized to `nullptr`, there is no guarantee it will always work.

I do not know how to properly use pointers, I will try to fix the initilazations thing.
for initialization will `IRsend *Plugin_035_irSender=nullptr;` do?

TD-er commented:
Yep, that should at least initialize it to 0.
As far as I can see both plugins do not have a `PLUGIN_EXIT` case where these pointers will be cleared, but that's something a lot of plugins lack. So it may still lead to some memory usage left when deleting one of these plugins.
But that's not something that will be done a lot.

jimmys01 committed:
[IR TX RX] Small fixes

jimmys01 committed:
[IR TX RAW] Uncomment code that was commented by mistake

jimmys01 committed:
[IR TX] Set Coolix to default to 1 repeat

Newest IR Library commit, fixes coolix protocol repeating the signal by default

jimmys01 commented:
Ok thats it. No more commits for this merge request.

TD-er commented:
@jimmys01 
So this one should be merged?
